update model card for gpt-j (#40404)
huggingface/transformers: docs/source/en/model_doc/gptj.md

-->
*This model was released on 2021-06-04 and added to Hugging Face Transformers on 2021-08-31.*

<div style="float: right;">
    <div class="flex flex-wrap space-x-1">
        <img alt="PyTorch" src="https://img.shields.io/badge/PyTorch-DE3412?style=flat&logo=pytorch&logoColor=white">

>         <img alt="TensorFlow" src="https://img.shields.io/badge/TensorFlow-FF6F00?style=flat&logo=tensorflow&logoColor=white">

Reviewer: Remove the TensorFlow and Flax badges

        ">
        <img alt="FlashAttention" src="https://img.shields.io/badge/%E2%9A%A1%EF%B8%8E%20FlashAttention-eae0c8?style=flat">
    </div>
</div>

# GPT-J

> [GPT-J](https://huggingface.co/papers/2306.05431) is a GPT-2-like causal language model trained on [the Pile](https://pile.eleuther.ai/) dataset. It was trained efficiently by computing attention and feedforward neural networks in parallel and uses rotary position embeddings to better inject positional information. The model has 6 billion parameters, 28 transformer layers, 16 attention heads, and a context window size of 2048 tokens. Its intended use is for English text generation but may also be adapted for summarization, code generation, categorization, and other NLP tasks.

Reviewer: ```suggestion
[GPT-J](https://github.com/kingoflolz/mesh-transformer-jax) is a GPT-like model trained on [the Pile](https://pile.eleuther.ai/) dataset. It was trained with the Mesh Transformer JAX framework, a model parallelism scheme. This model efficiently computes attention and feedforward neural networks in parallel and uses rotary position embeddings to better inject positional information.
```

> You can find all the original [GPT-J] checkpoints under the [EleutherAI](https://huggingface.co/EleutherAI) organization.

Reviewer: ```suggestion
You can find all the original [GPT-J] checkpoints under the [EleutherAI](https://huggingface.co/EleutherAI/models?search=gpt-j) organization.
```

> [!TIP]
> This model was contributed by [Stella Biderman](https://huggingface.co/stellaathena).
> Click on the GPT-J models in the right sidebar for more examples of how to apply GPT-J to different language tasks.

> The example below demonstrates how to generate text with [`Pipeline`], the [`AutoModel`], and from the command line.

Reviewer: ```suggestion
The example below demonstrates how to generate text with [`Pipeline`], [`AutoModel`], and from the command line.
```

<hfoptoins id="usage">
<hfoption id="Pipeline">

```py
import torch
from transformers import pipeline

pipeline = pipeline(
    task="text-generation",
    model="EleutherAI/gpt-j-6B",
    torch_dtype=torch.float16,
    device=0
)
pipeline(
    "In a shocking finding, scientists discovered a herd of unicorns living in a remote, previously unexplored valley, in the Andes Mountains."
)
```

</hfoption>

> <hfoption id="AutoModel">

Reviewer: ```py
import torch
from transformers import AutoModelForCausalLM, AutoTokenizer, infer_device

device = infer_device()
model = AutoModelForCausalLM.from_pretrained("EleutherAI/gpt-j-6B", dtype=torch.float16, attn_implementation="flash_attention_2").to(device)
tokenizer = AutoTokenizer.from_pretrained("EleutherAI/gpt-j-6B")

prompt = (
    "In a shocking finding, scientists discovered a herd of unicorns living in a remote, "
    "previously unexplored valley, in the Andes Mountains. Even more surprising to the "
    "researchers was the fact that the unicorns spoke perfect English."
)

input_ids = tokenizer(prompt, return_tensors="pt").input_ids.to(model.device)

output = model.generate(
    input_ids,
    do_sample=True,
    temperature=0.9,
    max_length=100,
)
tokenizer.batch_decode(output)[0]
```

```py
from transformers import GPTJForCausalLM, AutoTokenizer
import torch

device = "cuda"
model = GPTJForCausalLM.from_pretrained("EleutherAI/gpt-j-6B", torch_dtype=torch.float16).to(device)
tokenizer = AutoTokenizer.from_pretrained("EleutherAI/gpt-j-6B")

prompt = (
    "In a shocking finding, scientists discovered a herd of unicorns living in a remote, "
    "previously unexplored valley, in the Andes Mountains. Even more surprising to the "
    "researchers was the fact that the unicorns spoke perfect English."
)

input_ids = tokenizer(prompt, return_tensors="pt").input_ids.to(device)

gen_tokens = model.generate(
    input_ids,
    do_sample=True,
    temperature=0.9,
    max_length=100,
)
gen_text = tokenizer.batch_decode(gen_tokens)[0]
```

</hfoption>
<hfoption id="transformers CLI">

```bash

> echo -e "In a shocking finding, scientists discovered a herd of unicorns living in a remote, previously unexplored valley, in the Andes Mountains." | transformers run 
>     --task text-generation --model EleutherAI/gpt-j-6B --device 0

Reviewer: ```suggestion
echo -e "In a shocking finding, scientists discovered a herd of unicorns living in a remote, previously unexplored valley, in the Andes Mountains." | transformers run --task text-generation --model EleutherAI/gpt-j-6B --device 0
```

```

</hfoption>
</hfoptions>

## Notes

> - GPT-J-6B is not intended for deployment without fine-tuning, supervision, and/or moderation. It may generate harmful or offensive text and is not suitable for human-facing interactions.
> 
> - GPT-J-6B has not been fine-tuned for downstream contexts such as writing genre prose or commercial chatbots. It does not respond to prompts like ChatGPT does.

Reviewer: ```suggestion
- Training requires at least 4x model size GPU memory even with mixed precision. Explore options such as DeepSpeed or use the original codebase to train and fine-tune the model on TPU and convert to Transformers for inference (see [here](https://github.com/kingoflolz/mesh-transformer-jax/blob/master/howto_finetune.md)).

- Although the embedding matrix is 50400, only 50257 entries are used. The extra tokens are added for TPU efficiency. To avoid a mismatch between embedding matrix size and vocab size, the GPT-J tokenizer contains 143 extra tokens (`<|extratoken_1|>... <|extratoken_143|>`.
```

> ## Resources

Reviewer: Reformat the `## Resources` section:

```md
- Blog on how to [Deploy GPT-J 6B for inference using Hugging Face Transformers and Amazon SageMaker](https://huggingface.co/blog/gptj-sagemaker).
- Blog on how to [Accelerate GPT-J inference with DeepSpeed-Inference on GPUs](https://www.philschmid.de/gptj-deepspeed-inference).
- Blog post introducing [GPT-J-6B: 6B JAX-Based Transformer](https://arankomatsuzaki.wordpress.com/2021/06/04/gpt-j/).
- Notebook for [GPT-J-6B Inference Demo](https://colab.research.google.com/github/kingoflolz/mesh-transformer-jax/blob/master/colab_demo.ipynb).
- Notebook demonstrating [Inference with GPT-J-6B](https://colab.research.google.com/github/NielsRogge/Transformers-Tutorials/blob/master/GPT-J-6B/Inference_with_GPT_J_6B.ipynb).
```
